src/audio/kpb.c

```diff
@@ -893,6 +897,130 @@ static int kpb_reset(struct comp_dev *dev)
 	return ret;
 }
 
+#ifdef KPB_HIFI3
+static void kpb_micselect_copy(struct comp_dev *dev, struct comp_buffer __sparse_cache *sink,
+			       struct comp_buffer __sparse_cache *source, size_t size,
+			       uint32_t channels)
+{
+	struct comp_data *kpb = comp_get_drvdata(dev);
+	struct audio_stream __sparse_cache *istream = &source->stream;
+	struct audio_stream __sparse_cache *ostream = &sink->stream;
+	uint16_t ch;
+	size_t i;
+
+	buffer_stream_invalidate(source, size);
+
+	if (kpb->config.sampling_width == IPC4_DEPTH_16BIT) {
+		const ae_int16 *in_ptr = (const ae_int16 *)istream->r_ptr;
+		ae_int16x4 d16 = AE_ZERO16();
+		const size_t in_offset = kpb->num_of_in_channels * sizeof(ae_int16);
+		const size_t out_offset = kpb->mic_sel_cnt * sizeof(ae_int16);
+		const size_t samples_per_chan = size / (sizeof(uint16_t) * kpb->mic_sel_cnt);
+		ae_int16 *buf = (ae_int16 *)(ostream->addr);
+		ae_int16 *buf_end = (ae_int16 *)(ostream->end_addr);
+		ae_int16 *out_ptr = (ae_int16 *)buf;
+
+		AE_SETCBEGIN0(buf);
+		AE_SETCEND0(buf_end);
+		for (ch = 0; ch < kpb->mic_sel_cnt; ch++) {
+			const ae_int16 *input_data = (const ae_int16 *)(in_ptr) + kpb->offsets[ch];
+
+			out_ptr = (ae_int16 *)ostream->w_ptr;
+			out_ptr += ch;
+			for (i = 0; i < samples_per_chan; i++) {
+				AE_L16_XP(d16, input_data, in_offset);
+				AE_S16_0_XC(d16, out_ptr, out_offset);
+			}
+		}
+	} else if (kpb->config.sampling_width == IPC4_DEPTH_32BIT) {
+		const ae_int32 *in_ptr = (const ae_int32 *)istream->r_ptr;
+		ae_int32x2 d32 = AE_ZERO32();
+		const size_t in_offset = kpb->num_of_in_channels * sizeof(ae_int32);
+		const size_t out_offset = kpb->mic_sel_cnt * sizeof(ae_int32);
+		const size_t samples_per_chan = size / (sizeof(uint32_t) * kpb->mic_sel_cnt);
+		ae_int32 *buf = (ae_int32 *)(ostream->addr);
+		ae_int32 *buf_end = (ae_int32 *)(ostream->end_addr);
+		ae_int32 *out_ptr = (ae_int32 *)buf;
+
+		AE_SETCBEGIN0(buf);
+		AE_SETCEND0(buf_end);
```

Comment:
can we just put "AE_SETCBEGIN0(ostream->addr); AE_SETCEND0(ostream->end_addr);" out of the if statement? and then the buf and buf_end is unused.